### PR TITLE
Add DI support to ValidatorTypeCache

### LINKED ISSUE #111: Support DI in validators
#### Is the feature request related to a problem

Validators are creating using `Activator.CreateInstance` and as such don't support DI (via ctor injection) of dependencies.

#### Describe the solution

Extend the `ValidatorTypeCache` to support DI scenarios. Rough notes, but I'd need to try if it'd work.
* Add an overload for `AddValidatorsFromAssembly` that takes a factory method to resolve a validator
** Instead of instantiating the types it finds upfront and storing them in the type cache, it'd store (still against each type to be validated) the type of the validator that was found, to be resolved instead via DI later.
** Stores the factory method for use later via `TryGetValidators`
* Update the `TryGetValidators` method to use that factory method if it's not null, else does normal behaviour to other type cache.

It's likely the factory method would need to have an argument of the graph context, so most of us could abuse the UserContext for resolving that in the right scope. That would probably mean a new overload for `TryGetValidators` that accepts the context. Haven't traced that back yet, but hopefully that'd be possible.

#### Describe alternatives considered

Potentially we could have a completely different `ValidatorTypeCache` implementation for DI that is swapped in? 

#### Additional context

I need this to support a validation scenario where I need to pull additional info async in the validator. I'm sure this will come up for others too eventually, e.g. validate email isn't used on createUser kinda mutation or something.

### PR DESCRIPTION
**Pull request by sungam3r**
fixes #111 

Fully backward compatible, I decided to get by with additional overloads to add support for DI.

Usage:
```c#
public void ConfigureServices(IServiceCollection services)
{
    services.AddValidatorsFromAssemblyContaining<Startup>();
}
```

```c#
var options = new ExecutionOptions
{
};
var cache = new ValidatorTypeCache(useDependencyInjection: true);
cache.AddValidatorsFromAssembly(typeof(Startup).Assembly);
options.UseFluentValidation(cache);
```

### CODE REVIEW

File: src/GraphQL.FluentValidation/ValidatorTypeCache.cs [hunk 1]
```diff
@@ -9,14 +9,39 @@
 namespace GraphQL.FluentValidation
 {
     /// <summary>
-    /// Static cache for all <see cref="IValidator"/>.
```
Comment (sungam3r) [inline src/GraphQL.FluentValidation/ValidatorTypeCache.cs:12]: This cache never was static :)

**Comment by benmccallum** [comment]
Nice one 🥇. All thumbs up from me. 

It's been so long since I was in this area doing a complex validator that needed this, but I think essentially the scenario I had was that I couldn't use constructor injection at all as the instantiation of the validators was done upfront and using Activator. 

I think that was an intentional design decision from @SimonCropp for performance issues as he'd thought you'd not want to be instantiating them all the time; but I'd found that the default configuration of FluentValidation sets them all up as transient lifetime anyway, so I think it's good to have the flexibillity to go either way if you want perf vs DX.

This is perfect as it allows you to instantiate them from your DI container so instead of reaching in to the validator's `RootContext` to grab the `UserContext` and then doing DI off that, I can just pop those deps in the ctor which is a more familiar pattern for most.

Thanks for working on this. I've been getting smashed at work so it's been hard to get much OS stuff done recently.

File: src/GraphQL.FluentValidation/ServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+using System;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// Extension methods for <see cref="IServiceCollection"/> to add fluent validators types.
+    /// </summary>
+    public static class ServiceCollectionExtensions
```
Comment (benmccallum) [inline src/GraphQL.FluentValidation/ServiceCollectionExtensions.cs:13]: Hmm, I wonder if the default here should be Transient, which aligns with the default lifetime in FluentValidation.
https://github.com/FluentValidation/FluentValidation/blob/de1400928b9517bedf3cb8936b72a3d0334d0180/src/FluentValidation.DependencyInjectionExtensions/ServiceCollectionExtensions.cs#L33

**Comment by benmccallum** [comment]
I'm guessing Simon might want some docs written up for this too, showing both options.

File: src/GraphQL.FluentValidation/ValidatorTypeCache.cs (hunk 1 above)
Comment (benmccallum) [reply src/GraphQL.FluentValidation/ValidatorTypeCache.cs:12]: It's a "per-execution-level" kinda cache isn't it. I wonder the comment could say something like that.

**Comment by sungam3r** [comment]
> I think that was an intentional design decision from @SimonCropp **for performance issues** as he'd thought you'd not want to be instantiating them all the time

No performance issues at all if configure validators as singletons.

File: src/GraphQL.FluentValidation/ServiceCollectionExtensions.cs (hunk 2 above)
Comment (sungam3r) [reply src/GraphQL.FluentValidation/ServiceCollectionExtensions.cs]: Let's do it easier - just use `FluentValidation.DependencyInjectionExtensions`.

Comment (benmccallum) [reply src/GraphQL.FluentValidation/ServiceCollectionExtensions.cs]: Yep, nice one.

**Comment by benmccallum** [comment]
> No performance issues at all if configure validators as singletons.

Exactly, can choose what you want.

**Pull request merged**